pkg/certifier/datadog/datadog.go

```diff
@@ -0,0 +1,219 @@
+	limiter := rate.NewLimiter(rate.Every(rateLimitInterval), rateLimit)
+	transport := clients.NewRateLimitedTransport(version.UATransport, limiter)
+	defaultClient := &http.Client{Transport: transport}
+
+	d := &datadogCertifier{
+		httpClient:    defaultClient,
+		assemblerFunc: assemblerFunc,
+	}
+
+	// apply options
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	if err := d.fetchManifests(); err != nil {
+		return nil, fmt.Errorf("failed to fetch DataDog manifests: %w", err)
+	}
+
+	return d, nil
+}
+
+func (d *datadogCertifier) fetchManifests() error {
+	npmResp, err := d.httpClient.Get(NPM_MANIFEST_URL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch NPM manifest: %w", err)
+	}
+	defer npmResp.Body.Close()
+
+	if err := json.NewDecoder(npmResp.Body).Decode(&d.npmData); err != nil {
+		return fmt.Errorf("failed to decode NPM manifest: %w", err)
+	}
+
+	pypiResp, err := d.httpClient.Get(PYPI_MANIFEST_URL)
+	if err != nil {
+		return fmt.Errorf("failed to fetch PyPI manifest: %w", err)
+	}
+	defer pypiResp.Body.Close()
+
+	if err := json.NewDecoder(pypiResp.Body).Decode(&d.pypiData); err != nil {
+		return fmt.Errorf("failed to decode PyPI manifest: %w", err)
+	}
+
+	return nil
+}
+
+func (d *datadogCertifier) CertifyComponent(ctx context.Context, rootComponent interface{}, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	packageNodes, ok := rootComponent.([]*root_package.PackageNode)
+	if !ok {
+		return ErrDataDogComponentTypeMismatch
+	}
+
+	predicates := &assembler.IngestPredicates{}
+	currentTime := time.Now().UTC()
+
+	for _, node := range packageNodes {
+		purl := node.Purl
+
+		pkgInput, err := helpers.PurlToPkg(purl)
+		if err != nil {
+			logger.Debugf("failed to parse purl '%s' into package: %v", purl, err)
+			continue
+		}
+
+		// determine which dataset to check based on package type
+		var maliciousVersions []string
+		switch pkgInput.Type {
+		case "npm":
+			fullName := pkgInput.Name
+			if pkgInput.Namespace != nil && *pkgInput.Namespace != "" {
+				namespace := strings.TrimPrefix(*pkgInput.Namespace, "@")
+				namespace = strings.TrimPrefix(namespace, "%40")
+				fullName = "@" + namespace + "/" + pkgInput.Name
+			}
+			v, found := d.npmData[fullName]
+			if !found {
+				continue
+			}
+			maliciousVersions = v
+		case "pypi":
+			v, found := d.pypiData[pkgInput.Name]
+			if !found {
+				continue
+			}
+			maliciousVersions = v
+		default:
+			logger.Debugf("Skipping package %s, not npm or pypi", purl)
+			continue
+		}
+
+		// if no versions specified in dataset, skip
+		if len(maliciousVersions) == 0 {
+			// package known but no malicious versions listed?
+			continue
+		}
+
+		// certify only if the package has a specified version and that exact version is known malicious
+		if pkgInput.Version == nil {
+			logger.Debugf("Package %s has no version specified, skipping...", purl)
+			continue
+		}
```

Comment:
Am I reading it correctly that we'll ignore things like "aiohtttps" (in PyPI) because it applies to all versions?

Reply:
I am not sure, I initially thought that an empty malicious version list meant no malicious versions found and thats what the code currently does, but it is instead the case that an empty list means all versions are known malicious? I couldn't find anything in the datadog malicious dataset repo to suggest this.

Reply:
I believe the empty list means all versions are known malicious. For example "aiohtttps" is a typosquat of the "aiohttps" (notice the third 't'). I opened DataDog/malicious-software-packages-dataset#135 to request clarification, but I'm confident in my interpretation.

Reply:
after further inspection, given the number of misspelled versions of popular packages observed, i'm also now leaning towards your conclusion. if these packages are already in the malicious dataset, that strongly suggests their malicious intent at the package level, even if a specific version isn’t listed.

Reply:
i've now changed the code to reflect this interpretation.

Reply:
This interpretation has been confirmed on the issue you posed, thanks @funnelfiasco !